### PR TITLE
macho: allow unaligned offsets in object files

### LINKED ISSUE #12387: stage2 unable to build stage3 on x86_64 macos: panic: incorrect alignment
Zig Version: `0.10.0-dev.3511+17511a48f`

```
andy@Andrews-Air-2:~/dev/zig/build$ stage2/bin/zig build -p stage3 -Denable-llvm -Denable-stage1
MachO Flush... thread 7320547 panic: incorrect alignment
/Users/andy/dev/zig/src/link/MachO/Object.zig:104:72: 0x1032c60bc in link.MachO.Object.parse (zig1)
                    @alignCast(@alignOf(macho.nlist_64), &self.contents[symtab.symoff]),
                                                                       ^
/Users/andy/dev/zig/src/link/MachO/Archive.zig:228:21: 0x1032b8f44 in link.MachO.Archive.parseObject (zig1)
    try object.parse(allocator, cpu_arch);
                    ^
/Users/andy/dev/zig/src/link/MachO.zig:2676:51: 0x103377880 in link.MachO.resolveSymbolsInArchives (zig1)
            const object = try archive.parseObject(gpa, cpu_arch, offsets.items[0]);
                                                  ^
/Users/andy/dev/zig/src/link/MachO.zig:1108:42: 0x1032440dd in link.MachO.linkOneShot (zig1)
        try self.resolveSymbolsInArchives();
                                         ^
/Users/andy/dev/zig/src/link/MachO.zig:414:45: 0x1032266d8 in link.MachO.flush (zig1)
        .one_shot => return self.linkOneShot(comp, prog_node),
                                            ^
/Users/andy/dev/zig/src/link.zig:641:72: 0x103223788 in link.File.flush (zig1)
            .macho => return @fieldParentPtr(MachO, "base", base).flush(comp, prog_node),
                                                                       ^
/Users/andy/dev/zig/src/Compilation.zig:2335:24: 0x1031bef48 in Compilation.flush (zig1)
    comp.bin_file.flush(comp, prog_node) catch |err| switch (err) {
                       ^
/Users/andy/dev/zig/src/Compilation.zig:2316:23: 0x1031b1e35 in Compilation.update (zig1)
        try comp.flush(main_progress_node);
                      ^
/Users/andy/dev/zig/src/main.zig:3327:20: 0x103149030 in main.updateModule (zig1)
    try comp.update();
                   ^
```

Same command works fine on my M1 mac. You need to check out the `stage3-default` branch (#12368) to replicate this. I'm using homebrew LLVM and `zig env` says

```
error: Zig was built/linked incorrectly: LLVM and Clang have separate copies of libc++
       If you are dynamically linking LLVM, make sure you dynamically link libc++ too
```

### PR DESCRIPTION
**Pull request by kubkon**
When parsing dylibs (binary files as opposed to text-based stubs), every `__LINKEDIT` data section has to be aligned to a minimum of `@alignOf(u64)` as otherwise `dyld` will refuse to load it, which means we don't have to relax the requirements for dylib parsing.

Closes #12387

### CODE REVIEW

**Comment by kubkon** [comment]
@andrewrk could you locally rebase #12368 on top of this branch, and verify this fixes #12387 for you? I'll also double check on my 10.15.7 macOS in the meantime.

**Comment by andrewrk** [comment]
> @andrewrk could you locally rebase #12368 on top of this branch, and verify this fixes #12387 for you? I'll also double check on my 10.15.7 macOS in the meantime.

Counter-proposal, let's fix it in master branch and then I will rebase #12368 when it is landed.

**Comment by kubkon** [comment]
> > @andrewrk could you locally rebase #12368 on top of this branch, and verify this fixes #12387 for you? I'll also double check on my 10.15.7 macOS in the meantime.
> 
> Counter-proposal, let's fix it in master branch and then I will rebase #12368 when it is landed.

That's what I meant actually. Just wanted for you to locally test it with `stage3-default` branch on your x86 macOS host that it actually did fix the issue :-)

**Review by andrewrk — APPROVED** [review]
Confirmed that it solves the issue for me locally :+1:

**Comment by kubkon** [comment]
> Confirmed that it solves the issue for me locally :+1: 

Excellent!

File: src/link/MachO/Object.zig [hunk 1]
```diff
@@ -99,10 +99,11 @@ pub fn parse(self: *Object, allocator: Allocator, cpu_arch: std.Target.Cpu.Arch)
             },
             .SYMTAB => {
                 const symtab = cmd.cast(macho.symtab_command).?;
-                self.in_symtab = @ptrCast(
-                    [*]const macho.nlist_64,
-                    @alignCast(@alignOf(macho.nlist_64), &self.contents[symtab.symoff]),
-                )[0..symtab.nsyms];
+                // Sadly, SYMTAB may be at an unaligned offset within the object file.
+                self.in_symtab = @alignCast(@alignOf(macho.nlist_64), @ptrCast(
```
Comment (andrewrk) [inline src/link/MachO/Object.zig:103]: Sorry I don't know how I missed this before but these `@alignCast` are doing the exact same assertion they were before. This patch effectively does nothing

**Comment by andrewrk** [comment]
btw could it be perhaps that the objects in question *are* aligned properly, however, the readToEndAlloc call is only using 8 bytes alignment instead of page alignment?

File: src/link/MachO/Object.zig (hunk 1 above)
Comment (kubkon) [reply src/link/MachO/Object.zig:103]: Hmm, I'm surprised as this fixed it for me locally. I'll re-do the tests then.

**Comment by kubkon** [comment]
As you discovered, it's not the problem with alignment of the entire object file, but with misaligned symtab that is not always aligned to 8 bytes for some reason.

**Pull request merged**